### PR TITLE
F252: zkatdlog cleanup - add comments and improve error handling

### PR DESCRIPTION
**Pull request by KElkhiyaoui**
Signed-off-by: Kaoutar <kao@zurich.ibm.com>

### CODE REVIEW

File: token/core/zkatdlog/nogh/auditor.go [hunk 1]
```diff
@@ -14,29 +15,42 @@ import (
 	api3 "github.com/hyperledger-labs/fabric-token-sdk/token/driver"
 )
 
+// AuditorCheck verifies if the passed tokenRequest matches the tokenRequestMetadata
 func (s *Service) AuditorCheck(tokenRequest *api3.TokenRequest, tokenRequestMetadata *api3.TokenRequestMetadata, txID string) error {
 	logger.Debugf("check token request validity...")
+	if s.TokenCommitmentLoader == nil {
```
Comment (adecaro) [inline token/core/zkatdlog/nogh/auditor.go:21]: no need for this check evey time. This is set at service construction time.

File: token/core/zkatdlog/nogh/auditor.go [hunk 2]
```diff
@@ -14,29 +15,42 @@ import (
...
+		return errors.New("failed to perform auditor check: nil token commitment loader")
+	}
 	var inputTokens [][]*token.Token
-	for _, transfer := range tokenRequestMetadata.Transfers {
+	for i, transfer := range tokenRequestMetadata.Transfers {
+		if &transfer == nil {
```
Comment (adecaro) [inline token/core/zkatdlog/nogh/auditor.go:26]: no need for this check, never nil

File: token/core/zkatdlog/nogh/auditor.go [hunk 3]
```diff
@@ -14,29 +15,42 @@ import (
...
 	des, err := s.Deserializer()
 	if err != nil {
-		return errors.WithMessagef(err, "failed getting deserializer")
+		return errors.WithMessagef(err, "failed getting deserializer for auditor check")
+	}
+	pp, err := s.PublicParams()
+	if err != nil {
+		return errors.Wrap(err, "failed to get public parameters for auditor check")
+	}
+	if pp == nil {
```
Comment (adecaro) [inline token/core/zkatdlog/nogh/auditor.go:44]: if no error, then pp is not nil. If it is nil, it is a programming error and panic is the right solution

File: token/core/zkatdlog/nogh/issuer.go [hunk 4]
```diff
@@ -59,8 +63,16 @@ func (s *Service) Issue(issuerIdentity view.Identity, typ string, values []uint6
 	return issue, infoRaws, fid, err
 }
 
+// VerifyIssue checks if the outputs of an IssueAction match the passed tokenInfos
 func (s *Service) VerifyIssue(ia driver.IssueAction, tokenInfos [][]byte) error {
 	action := ia.(*issue.IssueAction)
+	if action == nil {
```
Comment (adecaro) [inline token/core/zkatdlog/nogh/issuer.go:69]: can we check that ia is not nil and here that the cast was succefull?

File: token/core/zkatdlog/nogh/loaders.go [hunk 5]
```diff
@@ -48,12 +50,19 @@ type VaultTokenLoader struct {
...
+// matching the token identifiers, the corresponding zkatdlog tokens, the information of the
+// tokens in clear text and the identities of their owners
+// LoadToken returns an error in case of failure
 func (s *VaultTokenLoader) LoadTokens(ids []*token3.ID) ([]string, []*token.Token, []*token.TokenInformation, []view.Identity, error) {
 	var tokens []*token.Token
 	var inputIDs []string
 	var inputInf []*token.TokenInformation
 	var signerIds []view.Identity
-
+	if s.TokenVault == nil {
```
Comment (adecaro) [inline token/core/zkatdlog/nogh/loaders.go:62]: what about adding a constructor and making VaultTokenLoader instead? We can avoid this check every time.

File: token/core/zkatdlog/nogh/loaders.go [hunk 6]
```diff
@@ -118,8 +135,12 @@ func (s *VaultPublicParamsLoader) Load() (*crypto.PublicParams, error) {
 	return pp, nil
 }
 
+// ForceFetch retrieves the public parameters from the ledger
 func (s *VaultPublicParamsLoader) ForceFetch() (*crypto.PublicParams, error) {
 	logger.Debugf("force public parameters fetch")
+	if s.PublicParamsFetcher == nil {
```
Comment (adecaro) [inline token/core/zkatdlog/nogh/loaders.go:141]: what about adding a constructor and making VaultTokenLoader instead? We can avoid this check every time.

File: token/core/zkatdlog/nogh/loaders.go [hunk 7]
```diff
@@ -93,7 +102,15 @@ type VaultPublicParamsLoader struct {
 	PPLabel             string
 }
 
+// Load retrieves the public parameters. It first checks if the public parameters are cached.
+// If not Load fetches the parameters from the ledger
 func (s *VaultPublicParamsLoader) Load() (*crypto.PublicParams, error) {
+	if s.TokenVault == nil {
+		return nil, errors.New("can't load public parameters: please initialize token vault")
```
Comment (adecaro) [inline token/core/zkatdlog/nogh/loaders.go:109]: what about adding a constructor and making VaultTokenLoader instead? We can avoid this check every time.

File: token/core/zkatdlog/nogh/sender.go [hunk 8]
```diff
@@ -16,32 +16,41 @@ import (
 	"github.com/pkg/errors"
 )
 
+// Transfer returns a TransferAction as a function of the passed arguments
+// It also returns the corresponding TransferMetadata
 func (s *Service) Transfer(txID string, wallet driver.OwnerWallet, ids []*token3.ID, outputTokens []*token3.Token, opts *driver.TransferOptions) (driver.TransferAction, *driver.TransferMetadata, error) {
 	logger.Debugf("Prepare Transfer Action [%s,%v]", txID, ids)
-
+	if s.TokenLoader == nil {
```
Comment (adecaro) [inline token/core/zkatdlog/nogh/sender.go:23]: no need for this check evey time. This is set at service construction time.

File: token/core/zkatdlog/nogh/sender.go [hunk 9]
```diff
@@ -51,30 +60,36 @@ func (s *Service) Transfer(txID string, wallet driver.OwnerWallet, ids []*token3
...
+			return nil, nil, errors.WithMessage(err, "failed serializing token info for zkatdlog transfer action")
 		}
 		infoRaws = append(infoRaws, raw)
 	}
-
+	// audit info for receivers
 	var receiverAuditInfos [][]byte
-	for _, output := range outputTokens {
+	for i, output := range outputTokens {
+		if output.Owner == nil {
```
Comment (adecaro) [inline token/core/zkatdlog/nogh/sender.go:82]: and what if this is a redeem?

File: token/core/zkatdlog/nogh/sender.go [hunk 10]
```diff
@@ -108,32 +123,42 @@ func (s *Service) Transfer(txID string, wallet driver.OwnerWallet, ids []*token3
 	return transfer, metadata, nil
 }
 
+// VerifyTransfer checks the outputs in the TransferAction against the passed tokenInfos
 func (s *Service) VerifyTransfer(action driver.TransferAction, tokenInfos [][]byte) error {
 	tr, ok := action.(*transfer.TransferAction)
 	if !ok {
-		return errors.Errorf("expected *zkatdlog.Transfer")
+		return errors.New("expected *zkatdlog.Transfer")
```
Comment (adecaro) [inline token/core/zkatdlog/nogh/sender.go:130]: shall we check if the action is nil as well before the cast?

File: token/core/zkatdlog/nogh/service.go [hunk 11]
```diff
@@ -91,61 +93,88 @@ func NewTokenService(
...
+// ConfigManager returns the configuration manager associated with the service
 func (s *Service) ConfigManager() config.Manager {
 	return s.configManager
 }
 
-func (s *Service) PublicParams() *crypto.PublicParams {
-	return s.PPM.PublicParams()
+// PublicParams returns the public parameters associated with the service
+func (s *Service) PublicParams() (*crypto.PublicParams, error) {
+	if s.PPM == nil {
```
Comment (adecaro) [inline token/core/zkatdlog/nogh/service.go:157]: no need for this check evey time. This is set at service construction time.

File: token/core/zkatdlog/nogh/service.go [hunk 12]
```diff
@@ -91,61 +93,88 @@ func NewTokenService(
...
+		return nil, errors.New("failed to get public parameters: nil PublicParamsManager")
+	}
+	return s.PPM.PublicParams(), nil
 }
 
 func (s *Service) FetchPublicParams() error {
+	if s.PPM == nil {
```
Comment (adecaro) [inline token/core/zkatdlog/nogh/service.go:164]: no need for this check evey time. This is set at service construction time.

**Review by adecaro — CHANGES_REQUESTED** [review]

File: token/core/zkatdlog/nogh/auditor.go (hunk 1 above)
Comment (KElkhiyaoui) [reply token/core/zkatdlog/nogh/auditor.go:21]: OK

File: token/core/zkatdlog/nogh/auditor.go (hunk 2 above)
Comment (KElkhiyaoui) [reply token/core/zkatdlog/nogh/auditor.go:26]: OK

File: token/core/zkatdlog/nogh/auditor.go (hunk 3 above)
Comment (KElkhiyaoui) [reply token/core/zkatdlog/nogh/auditor.go:44]: OK

File: token/core/zkatdlog/nogh/issuer.go (hunk 4 above)
Comment (KElkhiyaoui) [reply token/core/zkatdlog/nogh/issuer.go:69]: OK

File: token/core/zkatdlog/nogh/sender.go (hunk 10 above)
Comment (KElkhiyaoui) [reply token/core/zkatdlog/nogh/sender.go:130]: OK

File: token/core/zkatdlog/nogh/service.go (hunk 11 above)
Comment (KElkhiyaoui) [reply token/core/zkatdlog/nogh/service.go:157]: OK

File: token/core/zkatdlog/nogh/loaders.go (hunk 6 above)
Comment (KElkhiyaoui) [reply token/core/zkatdlog/nogh/loaders.go:141]: For now I will remove this check.

File: token/core/zkatdlog/nogh/loaders.go (hunk 7 above)
Comment (KElkhiyaoui) [reply token/core/zkatdlog/nogh/loaders.go:109]: For now I will remove this check.

File: token/core/zkatdlog/nogh/sender.go (hunk 9 above)
Comment (KElkhiyaoui) [reply token/core/zkatdlog/nogh/sender.go:82]: even when it is redeem, owner is initialized.

**Review by adecaro — APPROVED** [review]

**Pull request merged**